### ISSUE TITLE
Causes communication errors

### ISSUE THREAD

**Issue by rafl** [kantlivelong/OctoPrint-SetAndWait#2]
#### What were you doing?

1. Run gcode using M190 with SetAndWait enabled

#### What did you expect to happen?

All gcode being executed with M190 substituted appropriately

#### What happened instead?

A communication error

#### Version of OctoPrint-SetAndWait

0.0.1

#### Operating System running OctoPrint

OctoPi 0.17.0

#### Printer model & used firmware incl. version

Prusa MK3S with MMU2S, firmware version 3.9.0 and 1.0.6, respectively

#### Link to octoprint.log with octoprnt.plugins.setandwait set to DEBUG

https://gist.github.com/rafl/fc671ea57e25db04e5bac8c9bc7ad355

**Comment by kantlivelong** [comment]
[Tx](https://gist.github.com/rafl/fc671ea57e25db04e5bac8c9bc7ad355#file-debug-gcode-L25) is for tool selection, correct? Are you getting prompted? 

Would you be able to test by specifying the tool number instead of a prompt?

Any idea what [Tc](https://gist.github.com/rafl/fc671ea57e25db04e5bac8c9bc7ad355#file-debug-gcode-L35) does?

**Comment by rafl** [comment]
Yeah, I'm being prompted by Tx on the printer LCD. Typically I use https://plugins.octoprint.org/plugins/mmu2filamentselect/, but had that disabled for the linked print. After selection, the specified filament is loaded up to the extruder gears.

Tc then subsequently loads the filament from the extruder gears all the way to the nozzle once it has reached its target temperature.

https://help.prusa3d.com/en/article/prusa-specific-g-codes_112173#t-codes has a tiny bit of documentation about it.

**Comment by kantlivelong** [comment]
I'm fairly certain that the issue has to do w/ `Tx`/`Tc` but not sure exactly why yet.

Would you be able to run the following tests and provide logs for each:
1. Going back to my previous reply could you try setting the tool / filament in the gcode file instead of using a prompt?
2. Try the same test as originally posted without this plugin or mmu2filamentselect and without selecting any tool/filament option when prompted. Just let it sit there for a few minutes after temps have reached the set value.

Thanks

**Comment by rafl** [comment]
1. https://gist.github.com/rafl/983f6dd4a406e7be89ddadcba18ac694
2. https://gist.github.com/rafl/88292eff342fa3d7b025bf7eae896fce

For 2, note that there's no way out of [`choose_menu_P`](https://github.com/prusa3d/Prusa-Firmware/blob/MK3/Firmware/ultralcd.cpp#L6058) without clicking the button.

**Comment by kantlivelong** [comment]
Thanks a bunch for running those tests. I think I may have a fix but haven't had a chance yet to create a way to simulate it in the virtual printer. Would you mind giving this version a go?

https://github.com/kantlivelong/OctoPrint-SetAndWait/archive/prusa_Tx_fix.zip

**Comment by kantlivelong** [comment]
Aside from the fix above for the comm issues it appears that using `Tx`/`T?`/`Tc` will result in the current tool that is tracked by OctoPrint will become out of sync with the controller. The [mmu2filamentselect](https://plugins.octoprint.org/plugins/mmu2filamentselect/) plugin might be a good workaround as it handles the selection locally in OctoPrint. 

The good news is that this is being discussed and there may be a solution at some point.

**Comment by kantlivelong** [comment]
@rafl can you confirm if the comm issue is now resolved w/ the version I linked?

**Comment by noxhirsch** [comment]
@kantlivelong With the old version the printer (MK3S MMU2S single mode) always got emergency shutdown from Octoprint with M112 after heatup. With you fix linked above, the problem is gone for me 👍
